### ISSUE TITLE
batch_jacobian incorrect

### ISSUE THREAD

**Issue by gaebor** [tensorflow/tensorflow#38428]
**System information** 
- Have I written custom code: yes
- OS Platform and Distribution
   - Windows 10, Anaconda Python 3.7.6, tensorflow 2.1.0
   - Debian GNU/Linux 8.11 (jessie), Anaconda Python 3.7.3, tensorflow 2.0.0

**Standalone code to reproduce the issue** 
```python
import numpy, tensorflow
x = tensorflow.Variable([[1], [1]], dtype="float64")
with tensorflow.GradientTape(persistent=True) as t:
    with tensorflow.GradientTape() as tt:
        obj = x[0]**2 + x[1]**2 + x[0]*x[1]
    dx = tt.gradient(obj, x)
print(t.batch_jacobian(dx, x).numpy())
```
output is 
```python
[[[3.]]

 [[3.]]]
```
which is not `[jacobian(y[i], x[i]) for i in range(x.shape[0])]` as [stated here](https://www.tensorflow.org/api_docs/python/tf/GradientTape#batch_jacobian).

It is more like `[gradient(sum(y), x[i]) for i in range(x.shape[0])]`, same as gradient of gradient:
```python
import numpy, tensorflow
x = tensorflow.Variable([1, 1], dtype="float64")
with tensorflow.GradientTape(persistent=True) as t:
    with tensorflow.GradientTape() as tt:
        obj = x[0]**2 + x[1]**2 + x[0]*x[1]
    dx = tt.gradient(obj, x)
print(t.jacobian(dx, x).numpy())
print(t.gradient(dx, x).numpy())
```
output is 
```python
[[2. 1.]
 [1. 2.]]
[3. 3.]
```
which is perfectly fine

**Other info / logs**
* f(x,y) = x^2+y^2+x y
* df/dx = 2x+y
* df/dy = 2y+x
* ddf/dxdx = 2
* ddf/dxdy = 1
* ddf/dydy = 2

What I want is the _diag(Hessian)_ without calculating full Hessian.

**Comment by gadagashwini** [comment]
Could able to reproduce the issue with Tf2.2rc2 and Tf2.1.
Please find the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/eb4009fce6fa8027c150363d52b51ca0/untitled500.ipynb). Thanks

**Comment by agarwal-ashish** [comment]
As per documentation of batch_jacobian, the function requires the following:

"target: A tensor with rank 2 or higher and with shape [b, y1, ..., y_n]. target[i,...] should only depend on source[i,...]."

In your case, the dependence requirement is not met.

**Comment by archilk** [comment]
I ran into this same issue today and thought it was a bug, but then saw @agarwal-ashish 's comment of the requirement not met (which is stated in the batch_jacobian API) and indeed it was the case for me.

That said, I think the documentation in the "Advanced Autodiff" guide could be improved by mentioning this, because right now that guide says:

> Caution: tf.GradientTape.batch_jacobian only verifies that the first dimension of the source and target match. It doesn't check that the gradients are actually independent. It's up to you to make sure you only use batch_jacobian where it makes sense. For example, adding a tf.keras.layers.BatchNormalization destroys the independence, since it normalizes across the batch dimension.

Hence, it seems from this warning that batch_jacobian doesn't care about y[i] being dependent on only x[i], and would still work as stated.

This is just a soft suggestion though, maybe it was just me who misunderstood.
